cpp/roaring.hh

```diff
@@ -300,15 +310,27 @@ class Roaring {
      * sparse bitmaps).
      */
     static Roaring read(const char *buf, bool portable = true) {
-        Roaring ans(NULL);
-        if (portable)
-            ans.roaring = roaring_bitmap_portable_deserialize(buf);
-        else
-            ans.roaring = roaring_bitmap_deserialize(buf);
-        if (ans.roaring == NULL) {
-            throw std::runtime_error("failed memory alloc while reading");
+        if (!portable) {
+
+            if (*(const unsigned char *)buf == SERIALIZATION_ARRAY_UINT32) {
+                /* This looks like a compressed set of uint32_t elements */
+                uint32_t card;
+                memcpy(&card, buf + 1, sizeof(uint32_t));
+                const uint32_t *elems = (const uint32_t *)(buf + 1 + sizeof(uint32_t));
+
+                return Roaring((size_t)card, elems);
+            } else if (buf[0] == SERIALIZATION_CONTAINER) {
+                buf += 1;
+            } else
+                throw std::runtime_error("bad serialization type designator in read input");
+        } {
+            Roaring ans;
+            bool is_ok = ra_portable_deserialize(&ans.roaring.high_low_container, buf);
+            if(!is_ok) {
+                throw std::runtime_error("failed memory alloc while reading");
+            }
+            return ans;
```

Comment:
this function is a bit less elegant and duplicates some of the code from the C source files but I think it is a fair trade for not perturbing the C at all